### ISSUE TITLE
Internal error: lookup for ComparisonExpr ((@2)[float] != ('NaN')[string])[bool]'s CmpOp failed

### ISSUE THREAD

**Issue by mrigger** [cockroachdb/cockroach#46914]
Consider the following statements:

```sql
CREATE TABLE t0(c0 INT);
SELECT * FROM t0 GROUP BY t0.c0 HAVING MIN((CASE WHEN false THEN NULL END):::FLOAT) IS NOT NAN; -- internal error: lookup for ComparisonExpr ((@2)[float] != ('NaN')[string])[bool]'s CmpOp failed
```
Unexpectedly, the `SELECT` results in an internal error with the following stack trace:
```
ERROR: internal error: lookup for ComparisonExpr ((@2)[float] != ('NaN')[string])[bool]'s CmpOp failed
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/sem/tree/expr.go:526: memoizeFn()
github.com/cockroachdb/cockroach/pkg/sql/sem/tree/expr.go:440: NewTypedComparisonExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar.go:228: buildComparison()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar.go:91: buildScalar()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar.go:206: buildBoolean()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar.go:91: buildScalar()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar.go:175: buildBoolean()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar.go:91: buildScalar()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:542: buildSelect()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:196: buildRelational()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:579: buildProject()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:199: buildRelational()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:142: build()
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:111: Build()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:178: makeOptimizerPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:796: makeExecPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:686: dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:476: execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:95: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1368: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1297: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:476: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:594: func1()
runtime/asm_amd64.s:1357: goexit()
```
I found this issue based on commit 51e8960053a8224856234ffc1fb610c3843f7d7d.

**Comment by asubiotto** [comment]
Looks like an issue in the `execbuilder`.

**Comment by RaduBerinde** [comment]
I spent some time on this. It's pretty gnarly. The core problem is that we are type-checking the argument to `MIN()` twice. Once just by itself when we are creating the aggregation argument (in `optbuilder/scope.replaceAggregate`) and once when we are typing the outer expression. The first time we typecheck the argument, we set the type inside `CastExpr` and elide the `TypeAnnotation`; the second time around, the annotation is not there anymore and we get a string.

This is an interaction of a few problems that we will need to address holistically at some point (including my favorite: modifying the AST in-place). For now, perhaps the right thing to do is to rewrite the expression and replace the aggregate argument with a variable (with the type we already resolved).

@rytaft I am putting this on your plate, but it is low priority (it's a fairly contrived corner case).
